pkg/serviceaccounts/oauthclient/oauthclientregistry.go

```diff
@@ -10,27 +12,150 @@ import (
+	OAuthRedirectModelAnnotationURIPortPrefix       = OAuthRedirectModelAnnotationURIPrefix + "port."
+	OAuthRedirectModelAnnotationURIPathPrefix       = OAuthRedirectModelAnnotationURIPrefix + "path."
+	OAuthRedirectModelAnnotationURIHostPrefix       = OAuthRedirectModelAnnotationURIPrefix + "host."
+	OAuthRedirectModelAnnotationResourceKindPrefix  = OAuthRedirectModelAnnotationResourcePrefix + "kind."
+	OAuthRedirectModelAnnotationResourceNamePrefix  = OAuthRedirectModelAnnotationResourcePrefix + "name."
+	OAuthRedirectModelAnnotationResourceGroupPrefix = OAuthRedirectModelAnnotationResourcePrefix + "group."
+
+	RouteKind = "Route"
 )
 
+var modelPrefixes = []string{
+	OAuthRedirectModelAnnotationURISchemePrefix,
+	OAuthRedirectModelAnnotationURIPortPrefix,
+	OAuthRedirectModelAnnotationURIPathPrefix,
+	OAuthRedirectModelAnnotationURIHostPrefix,
+	OAuthRedirectModelAnnotationResourceKindPrefix,
+	OAuthRedirectModelAnnotationResourceNamePrefix,
+	OAuthRedirectModelAnnotationResourceGroupPrefix,
+}
+
+type namesToObjMapperFunc func(namespace string, names sets.String) map[string]redirectURIList
+
+var routeGroupKind = unversioned.GroupKind{Group: routeapi.FutureGroupName, Kind: RouteKind}
+
 type saOAuthClientAdapter struct {
 	saClient     kclient.ServiceAccountsNamespacer
 	secretClient kclient.SecretsNamespacer
+	routeClient  osclient.RoutesNamespacer
 
 	delegate    oauthclient.Getter
 	grantMethod oauthapi.GrantHandlerType
 }
 
+type model struct {
+	scheme string
+	port   string
+	path   string
+	host   string
+
+	group string
+	kind  string
+	name  string
+}
+
+func (m *model) getGroupKind() unversioned.GroupKind {
+	return unversioned.GroupKind{Group: m.group, Kind: m.kind}
+}
+
+type modelList []model
+
+func (ml modelList) getNames() sets.String {
+	var data []string
+	for _, model := range ml {
+		if len(model.name) > 0 {
+			data = append(data, model.name)
+		}
+	}
+	return sets.NewString(data...)
+}
+
+func (ml modelList) getRedirectURIs(objMapper map[string]redirectURIList) redirectURIList {
+	var data redirectURIList
+	for _, m := range ml {
+		if uris, ok := objMapper[m.name]; ok {
+			for _, uri := range uris {
+				u := uri
+				u.merge(m)
+				data = append(data, u)
+			}
+		}
+	}
+	return data
+}
+
+type redirectURI struct {
+	scheme string
+	host   string
+	port   string
+	path   string
+}
+
+func (uri *redirectURI) String() string {
+	host := uri.host
+	if len(uri.port) > 0 {
+		host = net.JoinHostPort(host, uri.port)
+	}
+	return (&url.URL{Scheme: uri.scheme, Host: host, Path: uri.path}).String()
+}
+
+func (uri *redirectURI) isValid() bool {
+	return len(uri.scheme) > 0 && len(uri.host) > 0
+}
+
+type redirectURIList []redirectURI
+
+func (rl redirectURIList) extractValidRedirectURIStrings() []string {
+	var data []string
+	for _, u := range rl {
+		if u.isValid() {
+			data = append(data, u.String())
+		}
+	}
+	return data
+}
+
+func (uri *redirectURI) merge(m model) {
```

Comment:
Eh, `merge` makes more sense to me since I am merging all the useful information in `model` with the `redirectURI`.